src/Storages/Hive/StorageHive.cpp

```diff
@@ -0,0 +1,680 @@
+{
+    LOG_TRACE(log, "Append hive file {}", file_info.path);
+    String filename = getBaseName(file_info.path);
+    /// Skip temporary files starts with '.'
+    if (filename.find('.') == 0)
+        return {};
+
+    auto hive_file = createHiveFile(
+        format_name,
+        fields,
+        hdfs_namenode_url,
+        file_info.path,
+        file_info.last_modify_time,
+        file_info.size,
+        hivefile_name_types,
+        storage_settings,
+        context_);
+
+    /// Load file level minmax index and apply
+    const KeyCondition hivefile_key_condition(query_info, getContext(), hivefile_name_types.getNames(), hivefile_minmax_idx_expr);
+    if (hive_file->hasMinMaxIndex())
+    {
+        hive_file->loadMinMaxIndex();
+        if (!hivefile_key_condition.checkInHyperrectangle(hive_file->getMinMaxIndex()->hyperrectangle, hivefile_name_types.getTypes())
+                 .can_be_true)
+        {
+            LOG_TRACE(log, "Skip hive file {} by index {}", hive_file->getPath(), hive_file->describeMinMaxIndex(hive_file->getMinMaxIndex()));
+            return {};
+        }
+    }
+
+    /// Load sub-file level minmax index and apply
+    if (hive_file->hasSubMinMaxIndex())
+    {
+        std::set<int> skip_splits;
+        hive_file->loadSubMinMaxIndex();
+        const auto & sub_minmax_idxes = hive_file->getSubMinMaxIndexes();
+        for (size_t i = 0; i < sub_minmax_idxes.size(); ++i)
+        {
+            if (!hivefile_key_condition.checkInHyperrectangle(sub_minmax_idxes[i]->hyperrectangle, hivefile_name_types.getTypes())
+                     .can_be_true)
+            {
+                LOG_TRACE(log, "Skip split {} of hive file {}", i, hive_file->getPath());
+                skip_splits.insert(i);
+            }
+        }
+        hive_file->setSkipSplits(skip_splits);
+    }
+    return hive_file;
+}
+
+Pipe StorageHive::read(
+    const Names & column_names,
+    const StorageMetadataPtr & metadata_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context_,
+    QueryProcessingStage::Enum /* processed_stage */,
+    size_t max_block_size,
+    unsigned num_streams)
+{
+    HDFSBuilderWrapper builder = createHDFSBuilder(hdfs_namenode_url, context_->getGlobalContext()->getConfigRef());
+    HDFSFSPtr fs = createHDFSFS(builder.get());
+    auto hive_metastore_client = HiveMetastoreClientFactory::instance().getOrCreate(hive_metastore_url, getContext());
+    auto hive_table_metadata = hive_metastore_client->getTableMetadata(hive_database, hive_table);
+
+    std::vector<Apache::Hadoop::Hive::Partition> partitions = hive_table_metadata->getPartitions();
+    /// Hive files to read
+    HiveFiles hive_files;
+    /// Mutext to protect hive_files, which maybe appended in multiple threads
+    std::mutex hive_files_mutex;
+
+    ThreadPool pool{num_streams};
+    if (!partitions.empty())
+    {
+        for (const auto & partition : partitions)
+        {
+            pool.scheduleOrThrowOnError([&]()
```

Comment:
How many partitions are there usually? 
If a LOT, then does it make sense not to check on each read this, but to have a background task, which collects this list of files. So this list can be a little outdated, but may be it's ok and we will just ignore non existing files. And for some period (until this background task updates the list) it can lack some files, but this might also be ok, we can add a setting which will define how often (like `update_period`) this list is updated by background thread.

Reply:
Partition numbers depends on users who creaeted and inserted into hive table.  
In `StorageHive::read`, we filter partitions by partition key conditions in query, and filter hive files or stripes/rowgroups by minmax indexes in ORC/Parquet  metadata(to be implemented).  That is to say, which hive files to be selected depend on not only we have what hive files in HDFS, but also where conditions in query. 

Above all, it not a proper to have a backgroud task updating hive file list. But I understand you intention for reuse some results of list directory, which we had implemented in `HiveTableMetadata`